### PR TITLE
`torch.tensordot`: performance improvements when contracting to a scalar.

### LINKED ISSUE #145731: [Performance] `tensordot` has substantial overhead
### 🐛 Describe the bug

`torch.tensordot(A, B)` is ~50% slower than `torch.dot(A.flatten(), B.flatten())` for matrices of size≤512 on GPU, and on CPU it is even slower than `torch.trace(A.T @ B)` until n=1024.

Image

Image

## Why this matters

A common mistake is to compute a quantity expressed in a math textbook / paper like $\text{trace}(A^T B)$ naively how it is written, which is $O(n^3)$, when it could be computed as $\langle A \mid B \rangle_{F} = \sum_{ij} A_{ij}B_{ij}$ which is $O(n^2)$. (The former computes all off-diagonal entries of the matrix product, which are irrelevant for the trace).

This would make a good lint rule, but the most natural implementation of the Frobenius inner product, which is `tensordot`, has this huge overhead and so should be currently avoided for small matrices.

## Expectation

- `tensordot(A, B)` should not be substantially slower than `torch.dot(A.flatten(), B.flatten())`.
- `tensordot(A, B)` should usually be faster than `torch.sum(A*B)`, as the latter cannot take advantage of fused multiply-add instructions.

## References:

- https://github.com/numpy/numpy/issues/25713
- https://github.com/astral-sh/ruff/issues/9664
- Code: https://gist.github.com/randolf-scholz/9b166ff6b112820d123037448369a0b7

### Versions

<details>

Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.31.2
Libc version: glibc-2.35

Python version: 3.12.8 (main, Dec  4 2024, 08:54:12) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-51-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 570.86.10
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
CPU family:                           6
Model:                                141
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          4800,0000
CPU min MHz:                          800,0000
BogoMIPS:                             4992.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             10 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.14.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.5.1
[pip3] torchinfo==1.8.0
[pip3] triton==3.1.0
[conda] Could not collect

</details>

cc @msaroufim @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by nikitaved**
As per title.
Fixes https://github.com/pytorch/pytorch/issues/145731

Touches only compute. The CPU overhead can potentially be further reduced.

Before:
```python
In [3]: n = 512

In [4]: A = torch.rand(n, n)

In [5]: B = torch.rand(n, n)

In [6]: %timeit torch.tensordot(A, B, [[0, 1], [0, 1]])
2.04 ms ± 70 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: %timeit torch.tensordot(A, B, [[0, 1], [1, 0]])
2.85 ms ± 191 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: %timeit torch.tensordot(A, B, [[1, 0], [0, 1]])
2.9 ms ± 133 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [9]: %timeit torch.tensordot(A, B, [[1, 0], [1, 0]])
4.07 ms ± 262 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

After
```python
In [2]: n = 512

In [3]: A = torch.rand(n, n)

In [4]: B = torch.rand(n, n)

In [5]: %timeit torch.tensordot(A, B, [[0, 1], [0, 1]])
30.7 µs ± 2.51 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [6]: %timeit torch.tensordot(A, B, [[0, 1], [1, 0]])
141 µs ± 6.52 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [7]: %timeit torch.tensordot(A, B, [[1, 0], [0, 1]])
142 µs ± 4.03 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: %timeit torch.tensordot(A, B, [[1, 0], [1, 0]])
62.8 µs ± 4.31 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by nikitaved** [comment]
The distributed issues are real - `aten.dot.default` is missing a sharding strategy. Shall it be added, or what would be the best way to fallback to the sharding-friendly implementation when in the distributed mode?

**Comment by bdhirsh** [comment]
@albanD who do you think would be a good reviewer for perf improvement PRs like this?

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by nikitaved** [comment]
Hey, @albanD , thank you for the review! The distributed failures are real. Do you mind if I xfail `tensordot` for now, and then implement a sharding strategy for `dot` in a follow-up?

**Comment by albanD** [comment]
breaking tensordot on DTensor doesn't sound right no? :p That might be very disruptive.
Is there an easy sharding strategy we can add here? Otherwise we might have to skip this fast path for subclass using at::isTensorSubclassLike(tensor) to check.

**Review by albanD — APPROVED** [review]
Thanks for taking the time to fix DTensor!

**Comment by nikitaved** [comment]
The Triton issues are unrelated and are present in master.

File: aten/src/ATen/native/mps/operations/Blas.mm [hunk 1]
```diff
@@ -51,14 +51,16 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
...
-
   using namespace mps;
   using CachedGraph = MPSBinaryCachedGraph;
 
   dot_check(self, other);
 
+  // (@nikitaved): MacOS13 does not support dot for int64, so we dispatch to mul + sum instead.
+  if (!is_macos_13_or_newer(MacOSVersion::MACOS_VER_14_0_PLUS) && self.scalar_type() == ScalarType::Long) {
+    return (self * other).sum();
+  }
```
Comment (nikitaved) [inline aten/src/ATen/native/mps/operations/Blas.mm:62]: @malfet , could you please tell whether you are fine with these changes?

Comment (nikitaved) [reply aten/src/ATen/native/mps/operations/Blas.mm]: Removing this change. Otherwise unexpected xfails/xpasses.

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -817,11 +819,40 @@ Tensor tensordot(const Tensor& input1, const Tensor& input2, IntArrayRef dims1,
...
+  if ((t1.device().type() == at::kMPS || t2.device().type() == at::kMPS) || size1 != 1 || size2 != 1) {
+    if (t1.device().type() == at::kMPS || t2.device().type() == at::kMPS) {
+      std::cout << "MPS: mm path" << std::endl;
+    }
+    // permute and reshape for matrix multiplication
+    t1 = t1.permute(p1).reshape_symint({size1, csize});
+    t2 = t2.permute(p2).reshape_symint({csize, size2});
+    // multiply and reshape to target size
+    return at::mm(t1, t2).reshape_symint(rsizes);
+  } else {
```
Comment (nikitaved) [inline aten/src/ATen/native/Linear.cpp:839]: Ok, this is really weird. MPS takes the right path - the gemm one. However, the consistency tests fail on grad tests because of the shape mismatch (cpu () and MPS (1, 1)). If, however, I revert the change to the original code (the same as in this path), the result is fine. Any ideas, @albanD , @malfet ?

**Comment by kwen2501** [comment]
@malfet @albanD I wonder if there is a way to unblock?

**Comment by nikitaved** [comment]
Alternatively, I could change the dispatcher destination, otherwise I have no the ability to debug MPS (and the issue could be in the dispatcher?).

**Comment by albanD** [comment]
If the new sample being added was broken on MPS before this change already. It's fine to keep it broken on MPS after this change.
But IIUC, your change should have no impact at all on MPS behavior right?

**Comment by nikitaved** [comment]
Yes, it should not. The sample iirc is old, however.

**Comment by ngimel** [comment]
The error is in the new path, which will produce 0d tensor no matter what (`return at::dot(t1.flatten(), t2.flatten());`), whereas the sample in question (and many others) wants (1,1), or any number of 1-sized dims really. You need to reshape your output. Side note: the fact that we don't have tests that would actually catch this without MPS is a shame, OpInfos are not all they are cracked up to be. one-off test again is in order.

**Comment by ngimel** [comment]
Another side note: pytorch docs for tensordot are absolutely useless, I could not parse the sum expression no matter how I tried. Numpy documentation though is better `The shape of the result consists of the non-contracted axes of the first tensor, followed by the non-contracted axes of the second.` https://numpy.org/doc/stable/reference/generated/numpy.tensordot.html, so when contracting 2 3d tensors over 2 dimensions, each of them has 1 non-contracted dimension, so output is 2d

**Comment by nikitaved** [comment]
Thank you, @ngimel , for the insight. Let us use NumPy then as the source of truth? Yeah, I also got confused by the sum semantics -- one other way to make it easier is to assume that each dimension contraction follows the rules of matrix multiplication. Regarding the consistency and the sample -- do you want it to be fixed here, or could that be done in a follow-up?

**Comment by ngimel** [comment]
Note, today's behavior is consistent with numpy documentation, and without your PR cuda and cpu produce 2d result, as expected, for this sample. So you need to do correct reshaping in your PR, and I'd prefer that you add tests for this in this PR also, the fact that our existing tests don't catch incorrectly-dimensioned output is pretty bad.

**Comment by nikitaved** [comment]
The shape issue is resolved, I indeed forgot to properly reshape as per computed result shape. I will add the test for that as well.

File: aten/src/ATen/native/Linear.cpp [hunk 3]
```diff
@@ -811,11 +812,35 @@ Tensor tensordot(const Tensor& input1, const Tensor& input2, IntArrayRef dims1,
...
+    t2 = t2.permute(p2);
+
+    if (t1.is_contiguous() && t2.is_contiguous()) {
+      // If t1 and t2 are both contiguous, then flatten is a view,
+      // then dot is the method of choice
+      return at::dot(t1.flatten(), t2.flatten()).reshape_symint(rsizes);
+    } else {
+      // Otherwise mul + sum can be faster as it avoids at most 2x contiguous() calls
+      // NOTE: t1.dtype == t2.dtype -- check above
+      return (t1.squeeze() * t2.squeeze()).sum(t1.scalar_type()).reshape_symint(rsizes);
```
Comment (ngimel) [inline aten/src/ATen/native/Linear.cpp:841]: why do you need dtype argument for `sum` here?

Comment (nikitaved) [reply aten/src/ATen/native/Linear.cpp:841]: int8 promotes to int64, and there is a test with int8

Comment (ngimel) [reply aten/src/ATen/native/Linear.cpp:841]: ah I see, I didn't realize we support tensordot for non-floating types.

**Comment by nikitaved** [comment]
@ngimel , could you please suggest a strategy for the testing? Do I have to test against a meta-fuction, or just some simple targeted test would do?

**Comment by ngimel** [comment]
Targeted test will do. Meta function is not there because tensordot is a composite op

**Comment by nikitaved** [comment]
Sorry, finally was able to find some time. @ngimel , could you please have a look?

**Review by ngimel — APPROVED** [review]
